[Instrumentation.Hangfire] Add metrics support (#3258)
open-telemetry/opentelemetry-dotnet-contrib: src/OpenTelemetry.Instrumentation.Hangfire/README.md

    .Build();
```

## Metrics

> This instrumentation library collects metrics following a POC/draft definition of workflow metrics defined as part of [semantic-conventions/#1688](https://github.com/open-telemetry/semantic-conventions/issues/1688). 

Reviewer: ```suggestion
This instrumentation library collects metrics following a POC/draft definition of workflow metrics defined as part of [semantic-conventions/#1688](https://github.com/open-telemetry/semantic-conventions/issues/1688).
```

> is not needed for counting completed executions.

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level        | Values                  |
|-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------|
| `workflow.task.name`          | string | Name of the task (Hangfire job method)        | Required                 | e.g., `MyJob.Execute`   |
| `workflow.execution.result`   | string | The result of executing the task              | Required                 | `success`, `failure`    |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`              |
| `error.type`                  | string | The type of error that occurred               | Conditionally Required¹  | Exception type name     |

¹ Required if and only if the task execution failed.

#### workflow.execution.duration

Duration of an execution grouped by task, state, and result. Records duration for different
execution phases:

> - **state=pending**: Time spent waiting in queue (only when `RecordQueueLatency` is enabled)

Reviewer: ```suggestion
execution phases:

- **state=pending**: Time spent waiting in queue (only when `RecordQueueLatency` is enabled)
```

> to support measuring different execution phases (pending vs executing). This attribute is
> being proposed for inclusion in the official semantic conventions.

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level        | Values                  |
|-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------|
| `workflow.task.name`          | string | Name of the task (Hangfire job method)        | Required                 | e.g., `MyJob.Execute`   |
| `workflow.execution.result`   | string | The result of executing the task              | Required                 | `success`, `failure`    |
| `workflow.execution.state`    | string | The execution phase being measured            | Required                 | `pending`, `executing`  |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`              |
| `error.type`                  | string | The type of error that occurred               | Conditionally Required¹  | Exception type name     |

> ¹ Required if and only if the task execution failed.

Reviewer: ```suggestion
[1]: Required if and only if the task execution failed.
```

| Units          | Instrument Type  | Value Type |
|----------------|------------------|------------|
| `{executions}` | UpDownCounter    | `Int64`    |

> [!NOTE]
> This metric tracks the current state of job executions. When a job transitions
> to a new state, the previous state is decremented and the new state is incremented.

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level        | Values                                    |
|-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------------------------|
| `workflow.task.name`          | string | Name of the task (Hangfire job method)        | Required                 | e.g., `MyJob.Execute`                     |
| `workflow.execution.state`    | string | Current state of the execution                | Required                 | `pending`, `executing`, `completed`       |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`                                |

> | `error.type`                  | string | The type of error that occurred               | Conditionally Required¹  | Exception type name                       |

Reviewer: ```suggestion
| `error.type`                  | string | The type of error that occurred               | Conditionally Required[1] | Exception type name                       |
```

#### workflow.outcome

The number of workflow instances which have been initiated. In Hangfire, this tracks
individual job completions.

| Units          | Instrument Type | Value Type |
|----------------|-----------------|------------|
| `{workflows}`  | Counter         | `Int64`    |

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level        | Values                  |
|-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------|
| `workflow.definition.name`    | string | Name of the workflow (Hangfire job method)    | Required                 | e.g., `MyJob.Execute`   |
| `workflow.result`             | string | The result of the workflow                    | Required                 | `success`, `failure`    |
| `workflow.trigger.type`       | string | Type of trigger that initiated the workflow   | Required                 | `api`, `cron`, `schedule` |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`              |

> | `error.type`                  | string | The type of error that occurred               | Conditionally Required¹  | Exception type name     |

Reviewer: ```suggestion
| `error.type`                  | string | The type of error that occurred               | Conditionally Required[1] | Exception type name     |
```

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level        | Values                                    |
|-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------------------------|
| `workflow.definition.name`    | string | Name of the workflow (Hangfire job method)    | Required                 | e.g., `MyJob.Execute`                     |
| `workflow.state`              | string | Current state of the workflow                 | Required                 | `pending`, `executing`, `completed`       |
| `workflow.trigger.type`       | string | Type of trigger that initiated the workflow   | Required                 | `api`, `cron`, `schedule`                 |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`                                |
| `error.type`                  | string | The type of error that occurred               | Conditionally Required¹  | Exception type name                       |

> ¹ Required if and only if the workflow execution failed.

Reviewer: ```suggestion
[1]: Required if and only if the workflow execution failed.
```

  - `workflow.state` tracks complete workflow lifecycle (used in `workflow.status`)
  - See state mapping tables above for details
- **error.type**: Set to the full type name of the exception when a job fails

#### Workflow vs Execution Metrics

The instrumentation distinguishes between **workflow-level** and **execution-level** metrics:

**Workflow-level metrics** (`workflow.outcome`, `workflow.status`):
- Track the complete lifecycle of jobs, including pre-execution states (e.g., scheduled)
- Use `workflow.definition.name`, `workflow.state`, and `workflow.trigger.type`
- Provide visibility into all jobs regardless of execution status

**Execution-level metrics** (`workflow.execution.*`):
- Track only jobs that have entered the execution pipeline (enqueued or later)
- Use `workflow.task.name` and `workflow.execution.state`
- Do NOT include `workflow.trigger.type` (trigger is a workflow-level concept)
- Provide detailed execution performance metrics

## References

* [OpenTelemetry Project](https://opentelemetry.io/)
* [Hangfire Project](https://www.hangfire.io/)

> * [OpenTelemetry Workflow Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/workflow/workflow-metrics.md)

Reviewer 2: I am curious how you defined metrics in this package. For sure, there is nothing under this link at the time of writhing this comment.

Author: Sorry, I didn't get it. Should I keep the worong link ?

Author: I used this document - https://github.com/thompson-tomo/semantic-conventions/blob/feature/%231688_Workflows/docs/workflow/workflow-metrics.md

to `workflow.execution.state="executing"` for actual execution time.

> [!WARNING]
> Enabling `RecordQueueLatency` requires an additional database query per job
> execution to retrieve the enqueue timestamp. In high-throughput scenarios,
> this may impact performance.
>
> [!NOTE]
> Add `using OpenTelemetry.Metrics;` to access
> `HangfireMetricsInstrumentationOptions`.

### Available Metrics

The following metrics are emitted by this instrumentation:

**Execution-level metrics** track jobs that have entered the execution
pipeline:

- `workflow.execution.outcome` - Counter for completed executions
- `workflow.execution.duration` - Histogram for execution duration (pending
  and executing phases)
- `workflow.execution.status` - UpDownCounter for current execution state
- `workflow.execution.errors` - Counter for execution errors

**Workflow-level metrics** track the complete job lifecycle:

- `workflow.outcome` - Counter for completed workflows
- `workflow.status` - UpDownCounter for current workflow state (including
  scheduled jobs)

---

#### workflow.execution.outcome

The number of task executions which have been initiated.

| Units          | Instrument Type | Value Type |

> |----------------|-----------------|------------|

Reviewer: ```suggestion
| -------------- | --------------- | ---------- |
```

|-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------|
| `workflow.task.name`          | string | Name of the task (Hangfire job method)        | Required                 | e.g., `MyJob.Execute`   |
| `workflow.execution.result`   | string | The result of executing the task              | Required                 | `success`, `failure`    |
| `workflow.execution.state`    | string | The execution phase being measured            | Required                 | `pending`, `executing`  |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`              |
| `error.type`                  | string | The type of error that occurred               | Conditionally Required[1]  | Exception type name     |

[1]: Required if and only if the task execution failed.

#### workflow.execution.status

The number of actively running tasks grouped by task and the current state.

| Units          | Instrument Type  | Value Type |

> |----------------|------------------|------------|

Reviewer: ```suggestion
| -------------- | ---------------- | ---------- |
```

| `{executions}` | UpDownCounter    | `Int64`    |

> [!NOTE]
> This metric tracks the current state of job executions. When a job
> transitions to a new state, the previous state is decremented and the new
> state is incremented.

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level        | Values                                    |

> |-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------------------------|

Reviewer: ```suggestion
| ----------------------------- | ------ | --------------------------------------------- | ------------------------ | ----------------------------------------- |
```

| `workflow.task.name`          | string | Name of the task (Hangfire job method)        | Required                 | e.g., `MyJob.Execute`                     |
| `workflow.execution.state`    | string | Current state of the execution                | Required                 | `pending`, `executing`, `completed`       |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`                                |
| `error.type`                  | string | The type of error that occurred               | Conditionally Required[1]  | Exception type name                       |

[1]: Required if and only if the task execution failed.

**Hangfire State Mapping:**

Hangfire job states are mapped to workflow semantic convention states as follows:

| Hangfire State                     | Workflow State |

> |------------------------------------|----------------|

Reviewer: ```suggestion
| ---------------------------------- | -------------- |
```

| Scheduled, Enqueued, Awaiting      | `pending`      |
| Processing                         | `executing`    |
| Succeeded, Failed, Deleted         | `completed`    |

#### workflow.execution.errors

The number of errors encountered in task executions.

| Units     | Instrument Type | Value Type |

> |-----------|-----------------|------------|

Reviewer: ```suggestion
| --------- | --------------- | ---------- |
```

| `{error}` | Counter         | `Int64`    |

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level | Values                  |
|-------------------------------|--------|-----------------------------------------------|-------------------|-------------------------|
| `error.type`                  | string | The type of error that occurred               | Required          | Exception type name     |
| `workflow.task.name`          | string | Name of the task (Hangfire job method)        | Required          | e.g., `MyJob.Execute`   |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended       | `hangfire`              |

#### workflow.outcome

The number of workflow instances which have been initiated. In Hangfire, this tracks
individual job completions.

| Units          | Instrument Type | Value Type |

> |----------------|-----------------|------------|

Reviewer: ```suggestion
| -------------- | --------------- | ---------- |
```

| `{workflows}`  | Counter         | `Int64`    |

**Attributes:**

| Attribute                     | Type   | Description                                   | Requirement Level        | Values                  |

> |-------------------------------|--------|-----------------------------------------------|--------------------------|-------------------------|

Reviewer: ```suggestion
| ----------------------------- | ------ | --------------------------------------------- | ------------------------ | ----------------------- |
```

| `workflow.definition.name`    | string | Name of the workflow (Hangfire job method)    | Required                 | e.g., `MyJob.Execute`                     |
| `workflow.state`              | string | Current state of the workflow                 | Required                 | `pending`, `executing`, `completed`       |
| `workflow.trigger.type`       | string | Type of trigger that initiated the workflow   | Required                 | `api`, `cron`, `schedule`                 |
| `workflow.platform.name`      | string | The workflow platform being used              | Recommended              | `hangfire`                                |
| `error.type`                  | string | The type of error that occurred               | Conditionally Required[1] | Exception type name                       |

[1]: Required if and only if the workflow execution failed.

**Hangfire State Mapping:**

Hangfire job states are mapped to workflow semantic convention states as follows:

| Hangfire State                     | Workflow State | Workflow Trigger Type           |

> |------------------------------------|----------------|---------------------------------|

Reviewer: ```suggestion
| ---------------------------------- | -------------- | ------------------------------- |
```
